sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala

```diff
@@ -312,6 +312,32 @@ object NestedColumnAliasing {
   }
 }
 
+object GeneratorUnrequiredChildrenPruning {
+  def unapply(plan: LogicalPlan): Option[LogicalPlan] = plan match {
+    case p@Project(_, g: Generate) =>
+      val requiredAttrs = p.references ++ g.generator.references
+      var pruned = false
+      val newChild = if (!g.child.outputSet.subsetOf(requiredAttrs)) {
+        pruned = true
+        Project(g.child.output.filter(requiredAttrs.contains), g.child)
+      } else {
+        g.child
+      }
+      val unrequired = g.generator.references -- p.references
+      val unrequiredIndices = newChild.output.zipWithIndex.filter(t => unrequired.contains(t._1))
+        .map(_._2)
+      if (unrequiredIndices.toSet != g.unrequiredChildIndex.toSet) {
+        pruned = true
+      }
+      if (pruned) {
+        Some(p.copy(child = g.copy(child = newChild, unrequiredChildIndex = unrequiredIndices)))
+      } else {
+        None
+      }
```

Comment:
There seems some logic change in the new code: https://www.diffchecker.com/wEi4OAYD, e.g., how `requiredAttrs` is obtained.

Is it possible to keep the change "mechanical", e.g.,

```
case p @ Project(_, g: Generate) if p.references != g.outputSet =>
val requiredAttrs = p.references -- g.producedAttributes ++ g.generator.references	
val newChild = prunedChild(g.child, requiredAttrs)	
val unrequired = g.generator.references -- p.references	
val unrequiredIndices = newChild.output.zipWithIndex.filter(t => unrequired.contains(t._1))	
        .map(_._2)	
val newP = if (newChild!=g.Child || unrequiredIndices.toSet != g.unrequiredChildIndex.toSet) {
    p.copy(child = g.copy(child = newChild, unrequiredChildIndex = unrequiredIndices))
} else {
    None
}

The only difference is the last line so that we still can enter the `GeneratorNestedColumnAliasing` branch if no rewrite happens?

Reply:
The only logic change is that I changed
```val requiredAttrs = p.references -- g.producedAttributes ++ g.generator.references```
to 
```val requiredAttrs = p.references ++ g.generator.references```
The other changes is due to `prunedChild` is defined in the `ColumnPruning` class so I just inlined its logic.
For the change to `requiredAttrs`, the reason is that I think the previous logic is incorrect, I don't think `g.producedAttributes` should be removed from the `requiredAttrs` list.

Reply:
Is it possible to not replicate the logic of `prunedChild`?  
`prunedChild` is used by every other branches. In case the logic in `prunedChild` evolve in the future, this call site will not get the benefit?

>> I don't think g.producedAttributes should be removed from the requiredAttrs list.

iiuc, `g.producedAttributes` shouldn't exist in `g.child.output`?  If so, why do we need to include them in `requiredAttrs`?

Reply:
> Is it possible to not replicate the logic of `prunedChild`?

Done.

> iiuc, `g.producedAttributes` shouldn't exist in `g.child.output`? If so, why do we need to include them in `requiredAttrs` for `g.child`?

Sorry let me rephrase myself. The previous implementation is not incorrect, but it is redundant IMO. I think `g. producedAttributes` should not exist in `g.child.output`, even before this rule. But I am fine to add it back as I think it does no harm.

Reply:
>> The previous implementation is not incorrect, but it is redundant IMO.

I think the meaning of `prunedChild` is a bit vague, particularly what `allReferences` mean. 
Therefore, "but it is redundant" is based on reading the content of `prunedChild`, rather than per the contract of `prunedChild`.

A cleaner version of it imo looks like the following:
```
def prunedChild(c: LogicalPlan, requiredAttributes: AttributeSet): LogicalPlan =
    require(requiredAttributes.subsetOf(c.outputSet))
    if (c.outputSet != requiredAttributes) {
      Project(requiredAttributes, c)
    } else {
      c
    }
```

I'm fine with go with your current code. We can update `prunedChild` in a followup PR -- maybe it'll fail some tests.

Reply:
Nit: it seems that we don't really need `var pruned`:

```
if (!newChild.fastEq(g.Child) || unrequiredIndices.toSet != g.unrequiredChildIndex.toSet) {
  p.copy(child = g.copy(child = newChild, unrequiredChildIndex = unrequiredIndices)) 
} else {
  None 
}
```

`||` can short circuit the second set equality check condition. But I think in reality, your code should also be fine. So it's your call.